### DISCUSSION TITLE
additionalProperties "sms" not allowed

### DISCUSSION THREAD

**Discussion by Anthony-J-Garot** [ory/kratos#2283] [Q&A] [answered]
TL;DR: Can't get sms: map to be recognized by kratos server.

I began with quickstart as a substrate. I then added quickstart-postgres.yml into the mix.  
Added totp and other functionality. So far ory kratos is working.

Now I am trying to get sms to work. 

My external SMS resource uses AWS SNS via an API Gateway and two lambda functions. The AWS sandbox works, i.e. I can fire off an SMS text to my phone using Postman/cURL.

So . . . I am unable to get sms configured in Ory Kratos. 

The error message in the server log is as noted in the subject (sensitive values masked with asterisks):

```
kratos_1                      | The configuration contains values or keys which are invalid:
kratos_1                      | courier: map[sms:map[from:+1844******* request_config:map[auth:map[config:map[password:YourPass user:YourUsername] type:basic_auth] body:file:///conf/sms-templates/aws.request.jsonnet header:map[Content-Type:application/x-www-form-urlencoded] method:POST url:https://************.execute-api.us-east-1.amazonaws.com/sendSmsMessage]] smtp:map[connection_uri:smtps://test:test@mailslurper:1025/?skip_ssl_verify=true from_address:no-reply@biolifesolutions.com] template_override_path:/conf/courier-templates]
kratos_1                      |          ^-- additionalProperties "sms" not allowed
```

PyCharm even complains when I add the sms: map into the kratos.yml file under the courier: label.

```
Schema validation: Property 'sms' is not allowed
```

So I upgraded my repo to the master branch and rebuilt. Same issue.

Digging a little further, the config.schema.json file that my kratos.yml file uses (v0.8.2-alpha.1) doesn't have the sms: definition.

`https://raw.githubusercontent.com/ory/kratos/v0.8.2-alpha.1/embedx/config.schema.json`

I did notice that the config.schema.json inside the repo *DOES* have the sms: map and seems to be described by [the online documentation](https://www.ory.sh/docs/kratos/concepts/email-sms).

`~/GitHub/ory/kratos/embedx/config.schema.json`

So . . . I'm wondering what is my next step to get sms: to work with ory kratos?

**Comment by zepatrik** [comment, marked as answer]
As you can see the feature commit is not yet released: https://github.com/ory/kratos/commit/687eca24aac7a7b89cc949693271343573107898
So to use this you have to build yourself from master. The pycharm validation will also work again with the release of that feature.

**Reply by Anthony-J-Garot** [reply]
As noted above, I built from master. I nuked everything in docker and rebuilt. I still get:

```
^-- additionalProperties "sms" not allowed
```

**Reply by aeneasr** [reply]
The tests pass, so it's most likely a problem with your local build. Maybe you're using the wrong binary, or the wrong docker image (the one from docker hub and not local)

**Reply by zepatrik** [reply]
It has to be a problem with your build @Anthony-J-Garot, I can add the SMS key without a problem if I run current master:
```
$ go run -tags sqlite . serve all -c ./contrib/quickstart/kratos/email-password/kratos.yml

The configuration contains values or keys which are invalid:
The configuration contains values or keys which are invalid:
courier.sms: map[auth:map[password:foobar user:asdf] enabled:false from:1.2345678e+07]
             ^-- validation failed

The configuration contains values or keys which are invalid:
courier.sms.from: 1.2345678e+07
                  ^-- expected string, but got number

The configuration contains values or keys which are invalid:
courier.sms: map[auth:map[password:foobar user:asdf] enabled:false from:1.2345678e+07]
             ^-- additionalProperties "auth" not allowed

The configuration contains values or keys which are invalid:
identity.schemas.0.url: <nil>
                        ^-- one or more required properties are missing

FATA[2022-03-04T10:00:33+01:00] Unable to instantiate configuration.          audience=application error=map[message:I[#] S[#] validation failed
  I[#/courier/sms] S[#/properties/courier/properties/sms] validation failed
    I[#/courier/sms/from] S[#/properties/courier/properties/sms/properties/from/type] expected string, but got number
    I[#/courier/sms] S[#/properties/courier/properties/sms/additionalProperties] additionalProperties "auth" not allowed
  I[#/identity/schemas/0] S[#/properties/identity/properties/schemas/items/required] missing properties: "url"] service_name=Ory Kratos service_version=master
```

There are validation errors, but others than yours.

**Reply by Anthony-J-Garot** [reply]
@aeneasr , your above response led me to get past the sms: map issue. I am now using ":latest" images by adding quickstart-latest.yml into the mix. This is the script I use to generate everything I want:

```
#!/bin/bash

# Created this bash script for the postgres line

docker build -f .docker/Dockerfile-build -t oryd/kratos:latest .

docker-compose \
	-f quickstart.yml \
	-f quickstart-standalone.yml \
	-f quickstart-latest.yml \
	-f quickstart-postgres.yml \
	up --build --force-recreate
```

@zepatrik , your above response was also helpful. I was indeed getting other issues that I was able to fix.
